### ISSUE TITLE
WinUI3 apps close after showing the splash screen when launched with an executionAlias

### ISSUE THREAD

**Issue by mrlacey** [microsoft/microsoft-ui-xaml#3441]
**Describe the bug**

If a WinUI3 app is launched with an [execution alias](https://docs.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-uap5-executionalias) the app starts but automatically closes after briefly showing the splash screen.

**Steps to reproduce the bug**

1. Create a new, empty WinUI3 desktop app from the VSIX template (version **3.0.0.2007130**)
2. Add the following to `package.appxmanifest` as a child of the `App` element.

```xml
        <Extensions>
            <uap5:Extension
                Category="windows.appExecutionAlias"
                Executable="$targetnametoken$.exe"
                EntryPoint="$targetentrypoint$">
                <uap5:AppExecutionAlias>
                    <uap5:ExecutionAlias Alias="MyTestApp.exe" />
                </uap5:AppExecutionAlias>
            </uap5:Extension>
        </Extensions>
```

(If using a UWP app set the EntryPoint as `{AppName}.App`)

3. Add the XML Namespace alias `xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"`
4. Run the app from VS to deploy it locally
5. Run the app from the command line `c:\> MyTestApp.exe`

**Expected behavior**

App should launch without issue

**Screenshots**

**Version Info**

NuGet package version: 
Microsoft.WinUI 3.0.0-preview2.200713.0

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| Yes | Yes |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2020 Update (19041)            |  Yes  |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            |  Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**

If in a WinUI3 UWP app, setting the Start action as "Do not launch, but debug my code when it starts" and adding a break point in the App Constructor , the break point is never hit. (I don't know how to test the equivalent in a desktop app--or if it's possible)

The error code/level after unsuccessfully launching an app from an alias is reported as **0**. This is a _success_ error code and so is misleading.

The same/equivalent functionality works fine for a pre-WinUI3/regular UWP app.

**Comment by mrlacey** [comment]
For the desktop app, I've also tried overriding the `OnActivated` event in the main app, and also defining `DISABLE_XAML_GENERATED_MAIN` and creating my own main method but this doesn't work with the executionAlias either. :(

**Comment by mrlacey** [comment]
Without this functionality it is not possible to pass command-line arguments to apps. 
While "Desktop" apps are being equated with "Win32" apps I feel the lack of command-line support is a big gap.

Currently this issue isn't even mentioned on the [roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md). With the GA release planned for H12021 and quarterly point releases after that, it looks as if command line support won't be available until at least Q4 2021.

This is a serious blocker for allowing WinUI3 apps from integrating with existing apps.

**Comment by marb2000** [comment]
I can still repro this in Preview 4.

**Comment by marb2000** [comment]
@mrlacey I made it work using the uap3:

```xml
<Package
  ...
  xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
  xmlns:desktop="http://schemas.microsoft.com/appx/manifest/desktop/windows10"
...
IgnorableNamespaces="uap rescap uap3">


    <Extensions>
        <uap3:Extension
              Category="windows.appExecutionAlias"
              EntryPoint="Windows.FullTrustApplication">
          <uap3:AppExecutionAlias>
            <desktop:ExecutionAlias Alias="MyTest.exe" />
          </uap3:AppExecutionAlias>
        </uap3:Extension>
      </Extensions>
```

Let me know if this works on your side.

**Comment by mrlacey** [comment]
@marb2000 I can confirm that **I can now get this working for both UWP and Desktop apps with Preview 4**. (My repro above  is incomplete 😢 and I can't find my original exploratory projects so I don't know if it was me or if it has been fixed.)

Both work with  `uap3&desktop` _and_ `uap5` aliases.


## Here's how I got it working in a WinUI3 UWP app.

- In `Package.appxmanifest`
    -  Add the alias `xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"`
    -  Add this below `</uap:VisualElements>`

```xml
    <Extensions>
        <uap5:Extension
                Category="windows.appExecutionAlias"
                Executable="App231.exe"
                EntryPoint="App231.App">
            <uap5:AppExecutionAlias>
                <uap5:ExecutionAlias Alias="MyTest.exe" />
            </uap5:AppExecutionAlias>
        </uap5:Extension>
    </Extensions>
```
OR use 
```xml
  xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
  xmlns:desktop="http://schemas.microsoft.com/appx/manifest/desktop/windows10"

   ...

    <Extensions>
        <uap3:Extension
                Category="windows.appExecutionAlias"
                Executable="App231.exe"
                EntryPoint="App231.App">
            <uap3:AppExecutionAlias>
                <desktop:ExecutionAlias Alias="MyTest.exe" />
            </uap3:AppExecutionAlias>
        </uap3:Extension>
    </Extensions>
```
(where "App231" is the name of the app/project and "MyTest.exe" is the alias to use to launch the app.)
Either set of extension (aliases) work the same.
- In `App.xaml.cs` add
```cs
    protected override void OnActivated(IActivatedEventArgs args)
    {
        base.OnActivated(args);

        Frame rootFrame = Window.Current.Content as Frame;

        if (rootFrame == null)
        {
            rootFrame = new Frame();
            rootFrame.NavigationFailed += OnNavigationFailed;
            Window.Current.Content = rootFrame;
        }

        rootFrame.Navigate(typeof(MainPage));
        Window.Current.Activate();
    }
```
 

## Here's how I got it working in a WinUI Desktop app:

- In `App.xaml.cs` of the main project, add:
```cs
    protected override void OnActivated(IActivatedEventArgs args)
    {
        base.OnActivated(args);

        if (args.Kind == ActivationKind.CommandLineLaunch)
        {
            m_window = new MainWindow();
            m_window.Activate();
        }
    }
```
- In `Package.appxmanifest` (of the "Package" project)
    -  Add the alias `xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"`
    -  Add this below `</uap:VisualElements>`
```xml
    <Extensions>
        <uap5:Extension
                Category="windows.appExecutionAlias"
                EntryPoint="Windows.FullTrustApplication">
            <uap5:AppExecutionAlias>
                <uap5:ExecutionAlias Alias="MyDesktopTest.exe" />
            </uap5:AppExecutionAlias>
        </uap5:Extension>
    </Extensions>
```
(where "MyDesktopTest.exe" is the alias to use to launch the app.)
This also works with `uap3` and `desktop` as above.


**Important points of note:**
- Using the method with `EntryPoint="Windows.FullTrustApplication"` requires the **restricted** `runFullTrust` capability which cannot be used with apps distributed through the store. If using/distributing the app separately this might not be an issue. The packaging project includes the `runFullTrust` capability by default in the (current) project templates. 
- I tested with WinUI3-**Preview4**, using **VS16.9.0 Preview 4.0**, on Windows Insider Preview **21296.1010**

**Comment by BorzillaR** [comment]
Hello!
My attempts to get these solutions to work in Project Reunion desktop app have failed.

Windows 20H2 (OS Build 19042.867)
Microsoft.ProjectReunion (0.5.0-prerelease)

**Comment by StephenLPeters** [comment]
@mrlacey does your solution work with ProjectReunion?

**Comment by mrlacey** [comment]
@StephenLPeters I've just run through [my steps above](https://github.com/microsoft/microsoft-ui-xaml/issues/3441#issuecomment-780583863) and **I can confirm that they work** for me with the ProjectReunion0.5-preview packages and templates.

It would be nice if WinUI had an automated test to verify this functionality, rather than rely on manual testing for regressions ;)

**Comment by StephenLPeters** [comment]
> It would be nice if WinUI had an automated test to verify this functionality, rather than rely on manual testing for regressions ;)

Yes, totally agree! @kmahone and @Austin-Lamb .  Thanks for confirming that :)

@borzillaR do you want to take a closer look or provide more info about your scenario?

**Comment by BorzillaR** [comment]
@StephenLPeters, I ran my app in the power shell. Now I launch it though the command prompt and it successfully showed. Should I use a different syntax to run my app through the power shell?

**Comment by mrlacey** [comment]
@BorzillaR you have to have run the app at elast once before being able to launch from the alias. Is that the issue you were having?

**Comment by BorzillaR** [comment]
@mrlacey, no, I am talking about another issue. 

After the first run I  do the following steps

Open a command line. 
Type MyDesktopTest.exe in the console.
The app starts successfully. 

Then open a Power Shell window. 
Type MyDesktopTest.exe there.
The app does not start at all.

Thats the issue.

**Comment by mrlacey** [comment]
@BorzillaR  this looks to be a PowerShell specific issue and not at all related to the issue originally reported here.

Just typing `MyDesktopTest.exe` would never be expected to work. PowerShell will not launch an executable when you pass it an executable name the way it does in other consoles/terminals.

In theory using the call operator (&) should work e.g. `& MyDesktopTest.exe` but this doesn't work and may be worth logging as a PowerShell issue.

However, **using Process-Start does work** to launch an application from an alias e.g. `Process-Start MyDesktopTest.exe`.

Hopefully this unblocks you.

**Comment by BorzillaR** [comment]
@mrlacey, I agree with you. This issue is not related to the original subject. The reason, I decided that it does not work, was I use the Power Shell console in my daily work. I could not imagine that the command line console would work better than Power Shell ))))

Anyway, it is strange that I can execute winver.exe from the Power Shell (whether by its name or by Start-Process) but I cannot execute MyDesktopTest.exe the same way. Another strange thing, when I type the command, which does not exist, I see a specific error in the response, but when I type MyDesktopTest.exe, there is no response. Looks like the app is launched but silently crashes.

![screen1](https://i.imgur.com/r828rDQ.gif)

In the command line, all work as expected.

![screen2](https://i.imgur.com/parnDaK.gif)

And thank you @mrlacey for the precise and clear instructions. It was very helpful for me.

**Comment by LaPeste** [comment]
> 
> 
> @BorzillaR you have to have run the app at elast once before being able to launch from the alias. Is that the issue you were having?

Could you elaborate a bit more on this?
